traffic_ops/traffic_ops_golang/deliveryservice/deliveryservicesv13.go

```diff
@@ -306,6 +312,88 @@ func createDefaultRegex(tx *sql.Tx, dsID int, xmlID string) error {
 	return nil
 }
 
+func parseOrgServerFQDN(orgServerFQDN string) (*string, *string, *string, error) {
```

Comment:
Can this use https://golang.org/pkg/net/url/#Parse ?

Reply:
It could, but I'd still have to parse the port out of `URL.Host` and make sure `URL.Scheme` is either http or https, and run through the rest of the URL struct to make sure things like Path, queryParams, and UserInfo don't exist. IMO doing it w/ the regex makes more sense since we're only interested in scheme, fqdn, and port.

Reply:
This doesn't validate the FQDN though. For example, `https://[:80` validates.

Even aside from the FQDN, I don't have any confidence that regex is correct. Regexes are always hard to read. I know the standard library is more code, but it's guaranteed to be correct, and it makes it easier for someone who comes along in the future to be certain the parsing is correct.

Reply:
Yes, I tried to make this regex as simple as possible, so it basically allows anything that isn't `:` to be parsed as the FQDN here. However, the `url.Parse` function doesn't validate it much either from what I can tell (aside from your example which returns an error because it's looking for a matching `]` as if it were enclosing an IPv6 address). But that's where the validation in `validateOrgServerFQDN` comes in. Anything parsed as the FQDN here must pass the `isHost` validation there.

And I disagree that regexes are always hard to read (especially this one). We're expecting a certain form of input pattern here, and I think a regex is the best tool for the job. If someone is having trouble understanding it, they can drop it into regex101.com and get a detailed description of the patterns and capture groups. The parentheses could even be removed to eliminate some of the obfuscation which makes it a relatively simple regex IMO.